File: cc/038.cc
// Project Euler Problem 038
//
// What is the max 1 to 9 pandigital that can be formed by
// multiplying a fixed number by 1, 2, 3, ... ?

#include <cstdio>
#include <cmath>
#include "mymath.h"

int cat_ints(int a, int b) {
  int p = log10(b) + 1;
  return pow(10.0, p) * a + b;
}

int main(int argc, const char* argv[]) {
  int max = 0;
  for (int i=0; i < 10000; i++) {
    int sum = 0;
    int bits = 1;  // pretend 0 is already set
    for (int j=1; j<6; j++) {
      int dj = digits(i*j);
      if (bits&dj) break;
      bits |= dj;
      sum = cat_ints(sum, i*j);
      if (bits==0x3ff && sum > max) {
          max = sum;
      }
    }
  }
  printf("%d\n", max);
}

File: go/problems/038.go
// Project Euler Problem 038
//
// What is the max 1 to 9 pandigital that can be formed by
// multiplying a fixed number by 1, 2, 3, ... ?

package problems

import (
	"fmt"
	"math"
)

func updateMaskAndNum(mask, fullNum, newNum int) (int, int) {
	pow10 := int(math.Pow10(int(math.Log10(float64(newNum)))))
	for newNum > 0 {
		digit := newNum / pow10
		bit := 1 << uint(digit)
		if mask&bit > 0 {
			return 0, 0
		}
		mask |= bit
		fullNum = fullNum*10 + digit
		newNum -= pow10 * digit
		pow10 /= 10
	}
	return mask, fullNum
}

func Problem038() string {
	max := 0
	for i := 1; i < 9876; i++ {
		mask := 1
		num := 0
		for n := 1; ; n++ {
			mask, num = updateMaskAndNum(mask, num, i*n)
			if mask == 0 {
				break
			}
			if mask == 0x3ff {
				if num > max {
					max = num
				}
				break
			}
		}
	}
	return fmt.Sprintf("%d", max)
}

func init() {
	Functions[38] = Problem038
}
